### PR TITLE
 Configuration Management via MonitoringContext

### LINKED ISSUE #5921: Unified Configuration Management System
## Feature
Currently the framework lacks a unified configuration management system. Many of our layers take an options bag which allows some level of configuration however this system is lacking, as it often requires a code change to change configuration values. This serves us ok for things like testing but works less well for deployed applications where ideally config is easier to change than code.

Some properties that such should provide:
-  Enable onetime overrides for dev testing debugging, and protect from unintentional spread.
- Configuration values should be immutable. once read in a session they should not change, as configs that can change increases complexity of consuming code.
- Developers should be able to easily hook into such a system to manage deployed resources
- Ideally configurations should continue to be strongly typed in the runtime
- The config system itself should be low overhead in terms of performance

### PR DESCRIPTION
**Pull request by anthony-murphy**
This change adds a simple configuration provider abstraction that can be injected via the loader. The configuration and logger are then available on an object currently called the MonitoringContext. To avoid needing to update all our api, we smuggle the MonitoringContext around via the logger.  After this initial change has had time to soak we can slowly move our apis over to leveraging MonitoringContexts all rather than logger. 

One are we have explicitly not tackled here is namespaces. this means we've gone with a flat structure or all config key look ups. In the future we may want to expand this introduce some kind of namespacing, but there are concerns we'll need to tackle around naming consistency. We face similar naming consistency issues with our child logger. Ideally whatever strategy we land on can help to solve the issues for both.

related to #5921

### CODE REVIEW

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [inline packages/utils/telemetry-utils/src/config.ts:287]: open to naming suggestions. Ideas i had were: ServicingContext, MeasureingContext

File: api-report/driver-base.api.md
Comment (andre4i) [inline api-report/driver-base.api.md:67]: I like this

File: packages/loader/container-loader/src/loader.ts
Comment (anthony-murphy) [inline packages/loader/container-loader/src/loader.ts:307]:
```suggestion
                [sessionStorageConfigProvider("Fluid").value, // Global override for all configs
```

Comment (anthony-murphy) [inline packages/loader/container-loader/src/loader.ts]: @andre4i you move session storage here, but i think it needs to be in `ConfigProvider.create` the driver, loader, and runtime all move at different deployment speeds, so it's likely that the loader will be behind so just setting it here will lead to other layers not having it if this change isn't there yet

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [inline packages/utils/telemetry-utils/src/config.ts:23]: @andre4i i wonder if it would be better to require the defaultValue, but allow undefined
like `getBoolean(name: string, defaultValue: boolean | undefined): boolean | undefined`

this would encourage developers to specify a default value, as they will need to explicitly specify undefined otherwise, rather than just leaving it off.

its a small change but i think we do want to encourage default values.

Comment (andre4i) [reply packages/utils/telemetry-utils/src/config.ts:23]: I agree.

File: packages/loader/container-loader/src/loader.ts
Comment (andre4i) [inline packages/loader/container-loader/src/loader.ts:307]: Should this namespace prefix be different? I think as is, the things we'd have to put inside `sessionStorageConfigProvider` would have to be 'Fluid.Fluid'

Comment (anthony-murphy) [reply packages/loader/container-loader/src/loader.ts:307]: so i've been looking at this and i think all the current namespace stuff is broken. the problem is similar to the above comment
> @andre4i you move session storage here, but i think it needs to be in `ConfigProvider.create` the driver, loader, and runtime all move at different deployment speeds, so it's likely that the loader will be behind so just setting it here will lead to other layers not having it if this change isn't there yet

basically, it's likely the loader won't provide config initially, so nested config won't get any prefix. this will lead to fully qualified config name changing over time, but i think we want stable names. I'm leaning toward config taking a fully qualified namespace, and just leaving it at that

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [inline packages/utils/telemetry-utils/src/config.ts:191]: @andre4i so i played around with typing for the default value and landed here,  we need to generic that infers from the parameter otherwise consumer always need to deal with undefined, even if the return can't be undefined because default is a real value. given this, i'm not sure default value is even worth it, and i'm not sure its worth it to introduce this complicated typeing.

example without the generic 
``
getBoolean(name: string, defaultValue: boolean | undefined): undefined | boolean 

// the typing here will be undefined | boolean eventhough it can't be undefined because my default is false
// so now i need to check, or use  ! or ?? anyway
const myconfig = getBoolean("name", false);
``

Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/config.ts:191]: i think i' prefer not to support a default, then consumers just do
``` typescript
const myconfig = getBoolean("name") ?? false;
```
and the typing is right without any generics or complicated typing

Comment (andre4i) [reply packages/utils/telemetry-utils/src/config.ts:191]: Yes, looking closely at how this API would be used currently in the codebase, I agree with you. We do stuff like `const myconfig = getBoolean("name", false) === true` anyway which makes the default value kind of cumbersome.

**Comment by markfields** [comment]
Alternate name idea that just came to me - DevopsContext?

File: api-report/driver-base.api.md
Comment (markfields) [inline api-report/driver-base.api.md:22]: @anthony-murphy, @andre4i - Alternate naming idea is **DevopsContext**.  Basically, stuff that is not part of the core business logic of Fluid, but necessary to leverage it in an app/service at scale.

MonitoringContext is fine too, it hints at the same thing, but I heard Tony wonder if there was a better name that was a more natural home for flighting/config, so wanted to share in case it resonated.

Comment (anthony-murphy) [reply api-report/driver-base.api.md:22]: I don't like Devops context, but i'm having trouble articulating why, so it could be a good option 😁Maybe you can help assuage my nebulous reservations.

I think the first is around messaging. Does devops as a term bring with it any connotations that we may not want? will it scare developers away thinking they need to follow devops, or know something about operations to use the fluid framework? will people confuse this with azure dev ops since both that and fluid framework are from Microsoft?

The second is the term itself. To me devops is an adjective, and i usually name with nouns and verbs, this is a thing, or this does a thing. So i'm not really sure what devops here does to modify the term context, it feels buzzwordy to me, and not really descriptive. 

I fully admit the monitoring also only really covers half the usage well, the logging, so its also not perfect, but it feel more neutral to me.

Also, these don't feel like they exactly express my reservations, but like i said, i having trouble articulating it, so open to being convinced.

Comment (markfields) [reply api-report/driver-base.api.md:22]: MonitoringContext is fine, as we just discussed.  I get the uneasiness about implying more than it really means :)

File: packages/utils/telemetry-utils/src/config.ts
Comment (andre4i) [inline packages/utils/telemetry-utils/src/config.ts:235]: I think there's some formatting inconsistencies in this file

File: packages/utils/telemetry-utils/src/index.ts
Comment (andre4i) [inline packages/utils/telemetry-utils/src/index.ts:22]: I'm curious why the explicit export here.

File: packages/utils/telemetry-utils/src/test/config.spec.ts
Comment (andre4i) [inline packages/utils/telemetry-utils/src/test/config.spec.ts:181]: How do you feel about removing this region altogether? I was thinking this would be useful but only as a guideline for how consumers of `SettingsProvider` are able to use our interfaces but I'm not entirely sure it belongs in the `main` branch

**Review by andre4i — APPROVED** [review]

File: packages/utils/telemetry-utils/src/index.ts
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/index.ts:22]: just minimize the exported surface area. there are things used in test that i don't want exported from the package itself

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/config.ts:235]: i'll take a look

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [reply packages/utils/telemetry-utils/src/config.ts:124]: Ok I don't think you're actually parsing "5" to 5, like I also mentioned above.

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [inline packages/utils/telemetry-utils/src/config.ts:228]: You don't need to say `T &`, it knows it's still `T`.

```suggestion
    obj: T,
): obj is MonitoringContext<T> {
```

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [inline packages/utils/telemetry-utils/src/config.ts:230]: No checks on the shape of `logger` prop?

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [reply packages/utils/telemetry-utils/src/config.ts:235]:
```suggestion
    if (loggerIsMonitoringContext(logger)) {
```

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [inline packages/utils/telemetry-utils/src/config.ts:234]: style nit:
```suggestion
    logger: T,
): MonitoringContext<T> {
```

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [reply packages/utils/telemetry-utils/src/config.ts:230]: Wait... I thought `maybeConfig` _is a_ logger in this case, not _has a_ logger...  Haven't looked at usages yet.  But that's what the types tell me, I think?

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [inline packages/utils/telemetry-utils/src/config.ts:242]: style nit:
```suggestion
    logger: T,
    ...configs: (IConfigProviderBase | undefined)[],
) {
```

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [reply packages/utils/telemetry-utils/src/config.ts:230]: Ohhh... when you mix it in you do `mc = logger; mc.logger = logger`. Wanna do `maybeConfig?.logger === maybeConfig`? 🤪 jk

**Review by markfields — COMMENTED** [review]
Phew I made it through `config.ts`...  I will have to come back to review the rest later (I guess after it's merged).

File: packages/utils/telemetry-utils/src/logger.ts
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:322]: style nit:
```suggestion
        properties: ITelemetryLoggerPropertyBags | undefined,
    ) {
```

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [reply packages/utils/telemetry-utils/src/config.ts:124]: Ok I guess this way someone can do `getString` to get the original string.  Makes sense.  But I still think it should be `raw: output`.

File: packages/utils/telemetry-utils/src/test/config.spec.ts
Comment (markfields) [inline packages/utils/telemetry-utils/src/test/config.spec.ts:61]: I'd drop the "stringAnd" cases from this test, and simply test `getString` on every single setting.  As-is, it's weird to have `number: "1"` above, and only be testing `getString` on those, when the arrays are also in there as strings.

File: packages/utils/telemetry-utils/src/test/config.spec.ts
Comment (markfields) [inline packages/utils/telemetry-utils/src/test/config.spec.ts:128]:
```suggestion
        assert.equal(config.getString("does not exist"), undefined);
```

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/config.ts:141]: The issue is we don't know intent. someone could have a config they think is a string, but parses to another type. they will call get a string, and expect the string they put in their config system. The goal here is to be as strongly typed as possible, but there are some ambiguous cases, and for those i've went with the approach that either is valid

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/config.ts:178]: there are places where we want a config provider, but are not sure it we have one, that leads to annoying code at the callsite like `const cp = maybeProvider == undefined ? new CachedConfigProvider() : CachedConfigProvider(maybeProvider)` it's even more annoying if you have multiple config providers. this makes that easy, as we just handled it here

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/config.ts:187]: this is a just an optimization, it won't work in all cases, and that's ok

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/config.ts:186]: it's probably not as necessary anymore, i uses to take a more complex type that included loggers, so this is from that originally

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/config.ts:230]: the comment references the behavior of this class, so you are correct

File: packages/utils/telemetry-utils/src/config.ts
Comment (anthony-murphy) [reply packages/utils/telemetry-utils/src/config.ts:231]: i early return where it is functionally necessary, i don't really see what`s inconsistent, here we need to cache the value, we could get and return it, or add a var, but that's lines of code that need to be downloaded on clients

File: packages/utils/telemetry-utils/src/test/config.spec.ts
Comment (andre4i) [reply packages/utils/telemetry-utils/src/test/config.spec.ts:61]: I'd leave these cases but probably a rename is in order to iron out any confusion. The problem was that the converter would have looked at the value (`"1"`), interpret it as a primitive type and return a number. This broke the scenario of trying to read a string which was actually the serialization of a number so every time you'd try to read a string, you'd get the default value because you weren't getting the expected type. When that was fixed, these test cases have been added.

Comment (markfields) [reply packages/utils/telemetry-utils/src/test/config.spec.ts:61]: How about this, for each case:
* getRawConfig should return the string value
* getFoo should return the value parsed as Foo
* getString should return the string value

Then remove the "stringAnd" cases as redundant.

File: packages/utils/telemetry-utils/src/config.ts
Comment (markfields) [reply packages/utils/telemetry-utils/src/config.ts:230]: Just giving feedback that the wording here was hard to understand at first.  Consider something like "cached config values are permanent" instead of "configs are immutable".

I'm bringing a different dialect to the table too, where "config" more naturally means what you're calling "config provider", and what you call "config" I would call "config value" or "setting".  So it's very subjective feedback :)

File: packages/drivers/driver-base/src/documentDeltaConnection.ts
Comment (markfields) [inline packages/drivers/driver-base/src/documentDeltaConnection.ts:295]: FC knows this is switching from "1" to "true"?

File: packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts
Comment (markfields) [inline packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts:133]:
```suggestion
        markUnreferencedNodes: boolean = this.mc.config.getBoolean("Fluid.Driver.Odsp.MarkUnreferencedNodes") ?? true,
```

File: packages/loader/container-loader/src/loader.ts
Comment (markfields) [inline packages/loader/container-loader/src/loader.ts:308]: Tangent:  I wish I could magically change this to just `"fluid"` namespace, the `":telemetry"` is just unnecessary

File: packages/loader/container-loader/src/loader.ts
Comment (markfields) [inline packages/loader/container-loader/src/loader.ts:311]:
```suggestion
            DebugLogger.mixinDebugLogger("fluid:telemetry", loaderProps.logger, { all: telemetryProps }),
            sessionStorageConfigProvider.value,
            loaderProps.configProvider,
        );
```

File: packages/loader/driver-utils/src/blobAggregationStorage.ts
Comment (markfields) [inline packages/loader/driver-utils/src/blobAggregationStorage.ts:175]: namespace?

Comment (markfields) [reply packages/loader/driver-utils/src/blobAggregationStorage.ts:175]: Also, I think this changes the semantics of the default value of `allowPacking`.  Previously, config was only checked if the arg was not supplied.  Now the config is the authority and the param is the fallback.

**Review by markfields — APPROVED** [review]
🎂 Really happy to see this get in!  LGTM once you've considered the comments I left, some may be changes you want to take.

**Pull request merged**
